### PR TITLE
Nimrod@GitHub description

### PR DESCRIPTION
**Pull request by NimrodGolan**

### CODE REVIEW

File: src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java [hunk 1]
```diff
@@ -205,6 +225,37 @@ private void handleIssueUpdateError(HttpClientErrorException e) {
...
+            commentFormat.append(issueStatus.getOpenFalsePositiveLinesAsADescription()).append("\n\n");
+        }
+
+        commentFormat.append("\n### **SUMMARY**\n\n");
+        if ((issueStatus.getTotalResolvedLinesFromResults() + issueStatus.getTotalResolvedFalsePositiveLines()) == 0) { // No Resolved vulnerabilities
+            commentFormat.append("Issue have total **").append(issueStatus.getTotalOpenLinesForIssueBeforeFixing()).append("** vulnerabilities left to be fix (Please scroll to the top for more information)");
+        } else {
+            commentFormat.append("- Total of vulnerabilities resolved on the last scan: **").append(issueStatus.getTotalResolvedLinesFromResults()).append("**");
+            commentFormat.append("\n- Total of vulnerabilities set as 'false positive' on the last scan: **").append(issueStatus.getTotalResolvedFalsePositiveLines()).append("**");
+            commentFormat.append("\n- Total of vulnerabilities left to fix for this issue: **").append(issueStatus.getTotalLinesToFixLeft()).append("**").append(" (Please scroll to the top for more information)");
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java]: Can you please shorten the lines to 120 characters?

File: src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java [hunk 2]
```diff
@@ -194,6 +197,23 @@ public Issue updateIssue(Issue issue, ScanResults.XIssue resultIssue, ScanReques
...
     }
 
+    private IssueStatus createNewIssueStatus(Issue issueBeforeFixing, ScanResults.XIssue resultIssue, com.checkmarx.flow.dto.github.Issue issueAfterFixing) {
+        Map<Integer, ScanResults.IssueDetails> sastFalsePositiveIssuesFromResult = getSASTFalsePositiveIssuesFromResult(resultIssue);
+        Map<String, String> sastResolvedIssuesFromResults = getSASTResolvedIssuesFromResults(issueBeforeFixing.getBody(), resultIssue);
+
+        IssueStatus issueStatus = IssueStatus.builder()
+                .sastResolvedIssuesFromResults(sastResolvedIssuesFromResults)
+                .openFalsePositiveLinesAsADescription(getNewFalsePositiveLines(sastFalsePositiveIssuesFromResult))
+                .totalOpenLinesForIssueBeforeFixing((extractGitHubIssueVulnerabilityCodeLines(issueAfterFixing.getBody()).size()))
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java]: Extracting local variables will make the code more readable.

File: src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java [hunk 3]
```diff
@@ -205,6 +225,37 @@ private void handleIssueUpdateError(HttpClientErrorException e) {
         }
     }
 
+    private void addCommentToAnUpdatedIssue(String issueUrl, IssueStatus issueStatus) {
+        StringBuilder commentFormat = setIssueUpdatedDescription(issueStatus);
+        this.addComment(issueUrl, commentFormat.toString());
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java]: is it commentFormat or the comment itself?

File: src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java [hunk 4]
```diff
@@ -185,7 +188,7 @@ public Issue updateIssue(Issue issue, ScanResults.XIssue resultIssue, ScanReques
         ResponseEntity<com.checkmarx.flow.dto.github.Issue> response;
         try {
             response = restTemplate.exchange(issue.getUrl(), HttpMethod.POST, httpEntity, com.checkmarx.flow.dto.github.Issue.class);
-            this.addComment(Objects.requireNonNull(response.getBody()).getUrl(),"Issue still exists. ");
+            addCommentToAnUpdatedIssue(Objects.requireNonNull(response.getBody()).getUrl(), createNewIssueStatus(issue, resultIssue, response.getBody()));
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java]: What about extracting the commenting logic into a separate class, e.g. CommentFormatter?

File: src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java [hunk 5]
```diff
@@ -205,6 +225,37 @@ private void handleIssueUpdateError(HttpClientErrorException e) {
         }
     }
 
+    private void addCommentToAnUpdatedIssue(String issueUrl, IssueStatus issueStatus) {
+        StringBuilder commentFormat = setIssueUpdatedDescription(issueStatus);
+        this.addComment(issueUrl, commentFormat.toString());
+    }
+
+    private StringBuilder setIssueUpdatedDescription(IssueStatus issueStatus) {
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java]: a better name would be something with "get", e.g. getUpdatedIssueComment

File: src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java [hunk 6]
```diff
@@ -205,6 +225,37 @@ private void handleIssueUpdateError(HttpClientErrorException e) {
...
+        StringBuilder commentFormat = new StringBuilder();
+        commentFormat.append("Issue still exists.\n");
+
+        if (!issueStatus.getSastResolvedIssuesFromResults().isEmpty()) {
+            commentFormat.append("The following code lines snippets were resolved from the issue:\n\n");
+            issueStatus.getSastResolvedIssuesFromResults().forEach((key, value) -> commentFormat.append("`Code line: ").append(key).append("`")
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java]: String.format will probably be more readable than append().

File: src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java [hunk 7]
```diff
@@ -205,6 +225,37 @@ private void handleIssueUpdateError(HttpClientErrorException e) {
...
+        }
+
+        if (!issueStatus.getOpenFalsePositiveLinesAsADescription().isEmpty()) {
+            commentFormat.append("\n");
+            commentFormat.append(issueStatus.getOpenFalsePositiveLinesAsADescription()).append("\n\n");
+        }
+
+        commentFormat.append("\n### **SUMMARY**\n\n");
+        if ((issueStatus.getTotalResolvedLinesFromResults() + issueStatus.getTotalResolvedFalsePositiveLines()) == 0) { // No Resolved vulnerabilities
+            commentFormat.append("Issue have total **").append(issueStatus.getTotalOpenLinesForIssueBeforeFixing()).append("** vulnerabilities left to be fix (Please scroll to the top for more information)");
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java]: "Issue has total"
"vulnerabilities" - did you mean findings?

File: src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java [hunk 8]
```diff
@@ -352,4 +403,87 @@ private static String extractTypeOfRelation(final String linkRelation) {
...
+
+    private Set<String> extractSASTResultCodeLines(ScanResults.XIssue resultIssue) {
+        return resultIssue.getDetails().entrySet()
+                .stream()
+                .map(e -> e.getKey().toString())
+                .collect(Collectors.toSet());
+    }
+
+    private Map<String, String> extractGitHubIssueVulnerabilityCodeSnippet(Set<String> resolvedIssueCodeLines, String issueBody) {
+        String codeSnippetForCodeLinePattern = "\\(Line #%s\\).*\\W\\`{3}\\W+(.*)(?=\\W+\\`{3})";
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java]: Please provide an example of such line in a comment.

File: src/main/java/com/checkmarx/flow/dto/github/IssueStatus.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+import java.util.Map;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+public class IssueStatus {
+
+    private String openFalsePositiveLinesAsADescription;
+    private Map<String, String> sastResolvedIssuesFromResults;
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/dto/github/IssueStatus.java]: Please provide a comment: what are the keys and values?

File: src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java (hunk 3 above)
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java]: addComment() gets the commentFormat as a string to execute the final comment on the issue

Comment (alex-ko-dev) [reply src/main/java/com/checkmarx/flow/custom/GitHubIssueTracker.java]: When I see xxxFormat, I expect a format string, something with %s etc.

**Pull request merged**